File: aoc-2025/cpp/src/day12.cpp
#include <cstdio>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>

const std::string DEFAULT_INPUT = "../input/day12.txt";
const std::string DEFAULT_OUTPUT = "output/day12.txt";

int part_one(const std::string &filename) {
  std::ifstream file(filename);
  if (!file.is_open())
    return -1;

  std::string line;

  for (int i = 0; i < 30; i++)
    std::getline(file, line);

  int ans = 0;

  while (std::getline(file, line)) {
    if (line.empty())
      continue;

    int w, h;
    sscanf(line.c_str(), "%dx%d:", &w, &h);

    int totalPieces = 0;
    std::istringstream iss(line.substr(line.find(':') + 1));

    int pieces;
    while (iss >> pieces)
      totalPieces += pieces;

    if (w * h >= totalPieces * 9)
      ans++;
  }
  return ans;
}

int part_two(const std::string &filename) {
  std::cout << "Merry Christmas 2025 !!!" << std::endl;
  return 0;
}

int main(int argc, char *argv[]) {
  std::string input_file = (argc > 1) ? argv[1] : DEFAULT_INPUT;
  std::cout << "Reading from: " << input_file << std::endl;

  auto p1 = part_one(input_file);
  auto p2 = part_two(input_file);

  std::cout << "Part 1: " << p1 << std::endl;
  std::cout << "Part 2: " << p2 << std::endl;

  std::ofstream oFile(DEFAULT_OUTPUT);
  if (oFile.is_open())
    oFile << p1 << "\n" << p2 << "\n";

  return 0;
}

File: aoc-2025/rust/src/bin/day12.rs
//! https://adventofcode.com/2025/day/12
use std::env;
use std::fs;
use std::path::Path;

const DEFAULT_INPUT: &str = "";
const DEFAULT_OUTPUT: &str = "";

fn part_one(content: &str) -> i32 {
    let mut ans = 0i32;

    for line in content.lines().skip(30) {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }

        let col_pos = match line.find(':') {
            Some(p) => p,
            None => continue,
        };

        let dims = &line[..col_pos];
        let mut dim_iter = dims.split('x');
        let w = match dim_iter.next().and_then(|s| s.trim().parse::<i32>().ok()) {
            Some(v) => v,
            None => continue,
        };
        let h = match dim_iter.next().and_then(|s| s.trim().parse::<i32>().ok()) {
            Some(v) => v,
            None => continue,
        };

        let pieces_str = &line[col_pos + 1..];
        let mut total_pieces = 0i32;
        for token in pieces_str.split_whitespace() {
            if let Ok(v) = token.parse::<i32>() {
                total_pieces += v;
            }
        }

        if w * h >= total_pieces * 9 {
            ans += 1;
        }
    }
    ans
}

fn part_two(_content: &str) -> i32 {
    println!("Merry Christmas 2025 !!!");
    0
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let input_path = args.get(1).map(|s| s.as_str()).unwrap_or(DEFAULT_INPUT);
    println!("Reading from: {}", input_path);

    let content = fs::read_to_string(input_path).expect("Could not read input file");

    let p1 = part_one(&content);
    let p2 = part_two(&content);

    println!("Part 1: {}", p1);
    println!("Part 2: {}", p2);

    if let Some(parent) = Path::new(DEFAULT_OUTPUT).parent() {
        let _ = fs::create_dir_all(parent);
    }
    let _ = fs::write(DEFAULT_OUTPUT, format!("{}\n{}\n", p1, p2));
}
